azext_edge/tests/edge/orchestration/test_upgrade2_unit.py

```diff
@@ -0,0 +1,413 @@
+# coding=utf-8
+# ----------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License file in the project root for license information.
+# ----------------------------------------------------------------------------------------------
+
+import json
+import re
+from typing import Dict, List, Optional, Tuple, TypeVar
+from unittest.mock import Mock
+
+import pytest
+import requests
+import responses
+from azure.cli.core.azclierror import ValidationError
+from azure.core.exceptions import HttpResponseError
+
+from azext_edge.edge.providers.orchestration.common import (
+    EXTENSION_ALIAS_TO_TYPE_MAP,
+    EXTENSION_MONIKER_TO_ALIAS_MAP,
+    EXTENSION_TYPE_ACS,
+    EXTENSION_TYPE_OPS,
+    EXTENSION_TYPE_OSM,
+    EXTENSION_TYPE_PLATFORM,
+    EXTENSION_TYPE_SSC,
+    EXTENSION_TYPE_TO_MONIKER_MAP,
+    ClusterConnectStatus,
+)
+from azext_edge.edge.providers.orchestration.targets import InitTargets
+from azext_edge.edge.util import parse_kvp_nargs
+
+from ...generators import generate_random_string
+from .resources.conftest import (
+    BASE_URL,
+    CLUSTER_EXTENSIONS_API_VERSION,
+    CLUSTER_EXTENSIONS_URL_MATCH_RE,
+    CONNECTED_CLUSTER_API_VERSION,
+    get_base_endpoint,
+    get_mock_resource,
+)
+from .resources.test_instances_unit import (
+    get_instance_endpoint,
+    get_mock_cl_record,
+    get_mock_instance_record,
+)
+
+T = TypeVar("T", bound="UpgradeScenario")
```

Comment:
prob use something more descriptive than just T (you can do US or UpScen or something if you want to keep it short)